src/transformers/trainer.py

```diff
@@ -2877,7 +2881,48 @@ def _load_rng_state(self, checkpoint):
                         "\nThis won't yield the same results as if the training had not been interrupted."
                     )
 
-    def _save_checkpoint(self, model, trial, metrics=None):
+    def _determine_best_metric(self, metrics, trial):
+        """
+        Determine if the model should be saved based on the evaluation metrics.
+        If args.metric_for_best_model is not set, the loss is used.
+
+        Returns:
+            bool: True if a new best metric was found, else False
+        """
+        new_best_metric = False
+
+        if self.args.metric_for_best_model is not None:
+            metric_to_check = self.args.metric_for_best_model
+
+            if not metric_to_check.startswith("eval_"):
+                metric_to_check = f"eval_{metric_to_check}"
+
+            try:
+                metric_value = metrics[metric_to_check]
+            except KeyError as exc:
+                raise KeyError(
+                    f"The `metric_for_best_model` training argument is set to '{metric_to_check}', which is not found in the evaluation metrics. "
+                    f"The available evaluation metrics are: {list(metrics.keys())}. Consider changing the `metric_for_best_model` via the TrainingArguments."
+                ) from exc
+
+            operator = np.greater if self.args.greater_is_better else np.less
+        else:
+            metric_value = metrics["eval_loss"]
+            operator = np.less
+
+        if self.state.best_metric is None or operator(metric_value, self.state.best_metric):
+            run_dir = self._get_output_dir(trial=trial)
+            checkpoint_folder = f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}"
+            output_dir = os.path.join(run_dir, checkpoint_folder)
+
+            self.state.best_metric = metric_value
+            self.state.best_model_checkpoint = output_dir
+
+            new_best_metric = True
+
+        return new_best_metric
```

Comment:
I prefer that we don't set a hardcoded metric `eval_loss` because it will be hard to find the issue. Instead, we should raise an error if the user do not set `metric_for_best_model` when using `SaveStrategy.BEST` or `load_best_model_at_end`.Hence, the last if block should only happen when inside`if self.args.metric_for_best_model is not None:`. 
LMK if this sounds good to you !

Reply:
Thanks for the suggestion, that sounds good to me. Just to make sure I understood correctly:

1. Throw an error if the user selects `SaveStrategy.BEST` or `load_best_model_at_end` but did not set a `metric_for_best_model`.
2. Remove the else block with the hard-coding of `metric_value`, hence only setting `new_best_metric = True` when `self.args.metric_for_best_model is not None`.

Let me know if there are any problems with my understanding. Otherwise, I'll go ahead and align my branch with `main` and implement the changes!

Reply:
that's right. Sorry for the delay !